File: scripts/test_normalize.js
// Test normalization logic for WhatsApp payloads
function normalizePayload(itemsJson) {
  var payload = itemsJson || {};
  try {
    if (payload && typeof payload.body === 'string') {
      try { payload = JSON.parse(payload.body); } catch (e) { /* ignore parse errors */ }
    } else if (payload && typeof payload.body === 'object' && payload.body !== null) {
      payload = payload.body;
    }
  } catch (e) { /* ignore */ }
  var text = '';
  var from = '';
  try {
    var entries = payload.entry;
    if (Array.isArray(entries) && entries.length > 0) {
      var changes = entries[0].changes;
      if (Array.isArray(changes) && changes.length > 0) {
        var value = changes[0].value;
        var messages = value && value.messages;
        if (Array.isArray(messages) && messages.length > 0) {
          var msg = messages[0];
          if (msg && msg.text && msg.text.body) text = msg.text.body;
          if (msg && msg.from) from = msg.from;
        }
      }
    }
  } catch (e) { /* ignore */ }
  return { user_text: text, from: from };
}

const sampleEntry = {
  entry: [
    {
      changes: [
        {
          value: {
            messages: [
              {
                from: '12345',
                text: { body: 'Halo dari test script' }
              }
            ]
          }
        }
      ]
    }
  ]
};

const direct = sampleEntry;
const wrappedObject = { body: sampleEntry };
const wrappedString = { body: JSON.stringify(sampleEntry) };

console.log('Direct payload:', normalizePayload(direct));
console.log('Wrapped object in body:', normalizePayload(wrappedObject));
console.log('Wrapped JSON string in body:', normalizePayload(wrappedString));

File: scripts/test_normalize.py
import json

def normalize(payload):
    p = payload or {}
    try:
        if isinstance(p.get('body'), str):
            try:
                p = json.loads(p['body'])
            except Exception:
                pass
        elif isinstance(p.get('body'), dict):
            p = p['body']
    except Exception:
        pass
    text = ''
    frm = ''
    try:
        entries = p.get('entry')
        if isinstance(entries, list) and len(entries) > 0:
            changes = entries[0].get('changes')
            if isinstance(changes, list) and len(changes) > 0:
                value = changes[0].get('value')
                messages = value and value.get('messages')
                if isinstance(messages, list) and len(messages) > 0:
                    msg = messages[0]
                    if msg and msg.get('text') and msg['text'].get('body'):
                        text = msg['text']['body']
                    if msg and msg.get('from'):
                        frm = msg['from']
    except Exception:
        pass
    return {'user_text': text, 'from': frm}

sample = {
  'entry': [
    {'changes': [
      {'value': {'messages': [{'from': '12345', 'text': {'body': 'Halo dari test script'}}]}}
    ]}
  ]
}

direct = sample
wrapped_obj = {'body': sample}
wrapped_str = {'body': json.dumps(sample)}

print('direct:', normalize(direct))
print('wrapped_obj:', normalize(wrapped_obj))
print('wrapped_str:', normalize(wrapped_str))
